src/Storages/HDFS/StorageHDFSCluster.cpp

```diff
@@ -25,13 +25,42 @@
 #include <Storages/IStorage.h>
 #include <Storages/SelectQueryInfo.h>
 #include <Storages/HDFS/HDFSCommon.h>
+#include <Storages/StorageDictionary.h>
 
 #include <memory>
 
 
 namespace DB
 {
 
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+static ASTPtr addColumnsStructureToQuery(const ASTPtr & query, const String & structure)
```

Comment:
Both the addColumnsStructureToQuery functions are very similar. If we take `max_arguments` as input, it can be 1 function and the error message for the exception can be appended by 's3 clustor /hdfs clustor' in the calling function.